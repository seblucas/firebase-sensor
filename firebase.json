{
  "hosting": {
    "public": "dist",
    "rewrites": [
      {
        "source": "**",
        "destination": "/index.html"
      }
    ],
    "headers": [
      {
        "source": "css/*.@(png|js|css|eot|otf|ttf|ttc|woff|woff2)",
        "headers": [
          {
            "key": "Cache-Control",
            "value": "max-age=604800"
          }
        ]
      },
      {
        "source": "js/*.@(png|js|css|eot|otf|ttf|ttc|woff|woff2)",
        "headers": [
          {
            "key": "Cache-Control",
            "value": "max-age=604800"
          }
        ]
      },
      {
        "source": "icon/*.@(png|js|css|eot|otf|ttf|ttc|woff|woff2)",
        "headers": [
          {
            "key": "Cache-Control",
            "value": "max-age=604800"
          }
        ]
      },
      {
        "source": "service-worker.js",
        "headers": [
          {
            "key": "Cache-Control",
            "value": "max-age=0"
          }
        ]
      },
      {
        "source": "favicon.ico",
        "headers": [
          {
            "key": "Cache-Control",
            "value": "max-age=604800"
          }
        ]
      }
    ]
  }
}
