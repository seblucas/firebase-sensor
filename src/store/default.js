import { getAuth, onAuthStateChanged, GoogleAuthProvider, signInWithPopup, signOut } from 'firebase/auth'
import { getDatabase, ref, get, onValue, remove, query, limitToLast } from 'firebase/database'

import ObjectToArray from '@/helper/object2array'
import devLog from '@/helper/devLog'

export const state = {
  firebaseApp: false,
  firebaseAuth: false,
  firebaseDatabase: false,
  user: null,
  rooms: false,
  categories: false,
  errors: []
}

export const mutations = {
  setUser (state, payload) {
    state.user = payload
  },
  setRooms (state, payload) {
    state.rooms = payload
  },
  setCategories (state, payload) {
    state.categories = payload
  },
  setErrors (state, payload) {
    state.errors = payload
  },
  setFirebaseApp (state, payload) {
    state.firebaseApp = payload
    state.firebaseAuth = getAuth()
    state.firebaseDatabase = getDatabase(state.firebaseApp)
  }
}

export const actions = {
  listenForAuthentication ({ dispatch, commit, getters }) {
    onAuthStateChanged(getters.firebaseAuth, (authData) => {
      if (authData) {
        devLog('Logged in as:', authData.uid)
        commit('setUser', authData)
        dispatch('loadRooms')
        dispatch('loadCategories')
        dispatch('loadErrors')
      } else {
        devLog('Logged out')
        commit('setUser', null)
        commit('setRooms', null)
        commit('setCategories', null)
        commit('setErrors', [])
      }
    })
  },
  login ({ getters }) {
    const provider = new GoogleAuthProvider()
    signInWithPopup(getters.firebaseAuth, provider).then(() => {
      // No need to do anything here it's handled by $onAuthStateChanged
    }).catch((error) => {
      console.error('Authentication failed:', error)
    })
  },
  logout ({ getters }) {
    signOut(getters.firebaseAuth).then(() => {
      devLog('Unauthentication completed')
    }).catch((error) => {
      console.error('Unauthentication failed:', error)
    })
  },
  loadRooms ({ commit, getters }) {
    get(ref(getters.firebaseDatabase, 'rooms')).then((newValue) => {
      commit('setRooms', newValue.val())
    }).catch((error) => {
      console.error(error)
    })
  },
  loadCategories ({ commit, getters }) {
    get(ref(getters.firebaseDatabase, 'readingCategories')).then((newValue) => {
      commit('setCategories', newValue.val())
    }).catch((error) => {
      console.error(error)
    })
  },
  loadErrors ({ commit, getters }) {
    const errors = ref(getters.firebaseDatabase, 'errors')
    onValue(errors, (newValue) => {
      commit('setErrors', ObjectToArray(newValue.val()))
    })
  },
  async loadDataFromFirebase ({ getters }, { roomId, startTimestamp, numberOfHours, categoryId }) {
    const queryReadings = query(ref(getters.firebaseDatabase, 'readings/' + roomId), limitToLast(4 * (numberOfHours + 1)))
    const newValue = await get(queryReadings)
    devLog(`Loaded from database for ${roomId}`)
    let basicArray = ObjectToArray(newValue.val())

    basicArray = basicArray.filter((item) => {
      // Remove too old readings and readings without the category
      return item.time > startTimestamp &&
        (!categoryId || Object.prototype.hasOwnProperty.call(item, categoryId))
    })
    return basicArray
  },
  async removeError ({ getters }, item) {
    await remove(ref(getters.firebaseDatabase, 'errors/' + item.id))
  }
}

export const getters = {
  firebaseAuth (state) {
    return state.firebaseAuth
  },
  firebaseApp (state) {
    return state.firebaseApp
  },
  firebaseDatabase (state) {
    return state.firebaseDatabase
  },
  user (state) {
    return state.user
  },
  rooms (state) {
    return state.rooms
  },
  categories (state) {
    return state.categories
  },
  errors (state) {
    return state.errors
  },
  categoriesSorted (state) {
    const basicArray = ObjectToArray(state.categories)
    return basicArray.sort((a, b) => a.order - b.order)
  }
}
