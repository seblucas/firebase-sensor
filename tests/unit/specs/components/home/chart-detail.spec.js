import { mockFirebaseResult } from '../../../mocks/firebase.mock'
import { shallowMount } from '@vue/test-utils'
import Vuex from 'vuex'
import ChartDetail from '@/components/home/chart-detail'
import genlocalVue from '../localVue'
import DevLog from '../../../mocks/devLog.mock'
import { mockStore } from '../../../mocks/store.mock'
import ObjectToArray from '@/helper/object2array'
import { FakeRooms, FakeCategories, FakeReadings, FakeRawReadings } from '../../../data/fake-data'

function genPropsData () {
  return {
    rooms: {
      room1: FakeRooms[0]
    },
    category: FakeCategories[0],
    numberOfHours: 24,
    readings: {
      room1: FakeReadings
    }
  }
}

function genWrapper (propsData) {
  const localVue = genlocalVue()
  const store = new Vuex.Store(mockStore)
  return shallowMount(ChartDetail, {
    localVue,
    store,
    mocks: { DevLog, ObjectToArray },
    attachTo: document.body,
    propsData
  })
}

describe('ChartDetail.vue', () => {
  it('shows a graph', () => {
    const wrapper = genWrapper(genPropsData())
    expect(wrapper.html().replace(/nv-edge-clip-[\d]*/g, '')
      .replace(/nv-chart-[\d]*/g, '')
      .replace(/id="[^"]*"/g, '')
      .replace(/clip-path="[^"]*"/g, '')
      // .replace(/d="[^"]*"/g, '')
      .replace(/class="[^"]*"/g, '')
      .replace(/transform="[^"]*"/g, '')).toMatchSnapshot()
  })
  it('load the data from Firebase if prop is empty', () => {
    mockFirebaseResult.mockReset()
    mockFirebaseResult.mockResolvedValue(FakeRawReadings)

    const propsData = genPropsData()
    propsData.readings = null
    const wrapper = genWrapper(propsData)
    expect(wrapper.vm.data).toHaveLength(1)
  })
  it('should reload the graph if the category is changed', async () => {
    const wrapper = genWrapper(genPropsData())
    const spy = jest.spyOn(wrapper.vm, 'loadDataAndGraph')
    wrapper.setProps({ category: FakeCategories[1] })
    await wrapper.vm.$nextTick()
    expect(spy).toHaveBeenCalled()
    jest.restoreAllMocks()
  })
  it('should reload the graph if the number of hours is changed', async () => {
    const wrapper = genWrapper(genPropsData())
    const spy = jest.spyOn(wrapper.vm, 'loadDataAndGraph')
    wrapper.setProps({ numberOfHours: 48 })
    await wrapper.vm.$nextTick()
    expect(spy).toHaveBeenCalled()
    jest.restoreAllMocks()
  })
  it.skip('should filter the data coming from Firebase to avoid having too old data',
    () => {})
  it.skip('should generate correct nvd3 parameter', () => {})
})
