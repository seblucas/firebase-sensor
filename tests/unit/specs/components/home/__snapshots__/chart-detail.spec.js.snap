// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChartDetail.vue shows a graph 1`] = `<div><canvas></canvas></div>`;
